### ISSUE TITLE
Replace deprecated gsl_sf_legendre_Plm_array functions

### ISSUE THREAD

**Issue by jobovy** [jobovy/galpy#289]
apparently with `gsl_sf_legendre_array` (see https://www.gnu.org/software/gsl/manual/html_node/Associated-Legendre-Polynomials-and-Spherical-Harmonics.html#Associated-Legendre-Polynomials-and-Spherical-Harmonics)

**Comment by SeaifanAladdin** [comment]
Switched to gsl_sf_legendre_array and gsl_sf_legendre_deriv_array

You'll need to update gsl to v2.2+

Currently running the unit tests. Will pull once it's done.

**Comment by jobovy** [comment]
Thanks for looking at this. Do ``gsl_sf_legendre_array`` and ``gsl_sf_legendre_deriv_array`` exist before v2.0? If not, I'm not sure how we should implement this, because most users are probably still on GSL v1.XX (this issue only came up today, even though these functions have been deprecated since October 2015).

**Comment by SeaifanAladdin** [comment]
I've taken a look and it looks like it only exists for v2.0+ 
I can also use gsl_sf_legendre_Plm instead, which computes the legendre for specified l and m. However, I'll have to compute it for each l and m without the use of recursion. It might be best if I just define my own recursive legendre function.

**Comment by jobovy** [comment]
Thanks. Let's just let this sit on the backburner for a bit and think it through; I don't think it's that urgent (even though I labeled this as "critical"!). Also imagine that you are busy with other things. 

I think we might be able to check which version of the gsl is installed using the preprocessor and then select the function based on that (or at runtime, but preprocessor would be better; these are the statements like ``#ifdef`` that you see in C code). At the very least, I think we can do this by defining a macro at compilation time that is set to the GSL main version number, so adding something like

``-D GSL_MAJOR_VERSION=%i" % (gsl_version[0])``

Maybe I'll start a branch and see whether this works.

We definitively not want to give up the recursive relationship, because otherwise the code will get very slow.

**Comment by jobovy** [comment]
For my own reference, here is the link to the C preprocessor options:

https://gcc.gnu.org/onlinedocs/gcc-4.4.3/gcc/Preprocessor-Options.html#Preprocessor-Options

**Comment by jobovy** [comment]
I did what I suggested in a branch ``gslv2``, see comparison here:

https://github.com/jobovy/galpy/compare/gslv2

However, I'm not sure whether the two GSL functions (in v1 and v2) do exactly the same thing and I don't currently have GSL v2 to test. If you have any insight on this @SeaifanAladdin, that would be much appreciated.

**Comment by SeaifanAladdin** [comment]
They both don't do the exact same thing, and they take different parameters. The function in v2 constructs the entire LxL matrix. This is great but there are two main problems. 

1. There are no non-deprecated functions that will recursively calculate the legendre polynomials for the axi symmetric case (where we set M=1). 

2. The functions from v2 fills an array P[L][M] with size L^2 - L*(L-1)/2 whereas in v1, it fills an array P[M][L] with size L^2. So not only are the indices flipped, but while I was able to create a triangular matrix with v1, v2 skips the m>l indices.

The solutions I was able to make are

1. For v2, I created a recursive function that calculates the Legendre polynomials for use in the axisymmetric case. For the L=100 case, I found the recursive function to be 4 times slower than `gsl_sf_legendre_deriv_array_e`, but when I set M=1, it becomes 15 times faster.

2. I've adjusted v1 such that it doesn't create a triangular matrix, and instead follows the convention of v2. So the size of P and dP is now L^2 -L(L-1)/2 in the non-axisymmetric case and L in the axisymmetric case. However, the indices of the arrays in v1 is still P[M][L] whereas v2 is P[L][M]. I handled that difference by calculating different indices for v1 and v2 at `computeNonAxi`.


I created a new branch and implemented everything. You can see the comparison here: [https://github.com/jobovy/galpy/compare/gslv2...SeaifanAladdin:gslv2](https://github.com/jobovy/galpy/compare/gslv2...SeaifanAladdin:gslv2)

I've confirmed this to work with gsl v2. Although I'm getting some failed tests with test_actionAngle.test_actionAngleTorus*. It looks like it has to do with MWPotential2014 and not the SCFPotential, but I'll take a look later. 

I haven't tested this with gsl v1 yet. I'll test it and compare execution speeds tomorrow.


Also, I used `gsl_sf_legendre_array_e` and `gsl_sf_legendre_deriv_array_e` because it has the needed (-1)^m term.

**Comment by SeaifanAladdin** [comment]
I've made new updates. Instead of using my recursive loops, I realize I could  use gsl_sf_legendre_Pl_array for the axisymmetric case. When M=1, gsl_sf_legendre_Plm_array reduced to gsl_sf_legendre_Pl_array.

So here is what I've done

    if M==1:

        gsl_sf_legendre_Pl_array

    else

        if v2:

            gsl_sf_legendre_array

        else

            gsl_sf_legendre_Plm_array


For a 100x100 matrix, gsl_sf_legendre_array is 1.75x faster than gsl_sf_legendre_Plm_array.

For a 100x1 matrix (axi-symmetric case), gsl_sf_legendre_Pl_array is about 1.45x faster than gsl_sf_legendre_Plm_array.


I have tested it with gsl v2, but haven't tested it with gsl v1.

**Comment by jobovy** [comment]
Thanks, this looks really good! Good point about the ``gsl_sf_legendre_Pl_array``.

I think we should make the ``if M == 1: ... else: ...`` an actual if-else statement and not use return inside the first if. You can just do
```
else{
#if GSL_MAJOR_VERSION == 2
    gsl_sf_legendre_array_e(GSL_SF_LEGENDRE_NONE,L - 1, x, -1, P_array);
#else
    int m;
    for (m = 1; m < M; m++)
    {
        gsl_sf_legendre_Plm_array(L - 1, m, x, P_array);
        P_array += L - m;
    }
#endif

and similar in the other similar if-else statement.

There's no way that you can just transpose the P[L][M] matrix in one of the versions such that they are the same later in the code? I guess that is difficult with the triangular matrix? Also, your code should never try to access the ``m>l`` elements, so is there any need for fancy indexing?

I'll try to setup the .travis.yml file such that it tests GSL v2 as well (currently it only uses v1, because that's what seems to be in apt-get).

**Comment by jobovy** [comment]
I have now setup the Travis builds in the ``gslv2`` branch such that they will run both GSL v1 and v2 (and the GSL v2 build appropriately fail for SCF). So if you merge my ``gslv2`` branch into yours and then open a galpy pull request, the new code should get tested.

**Comment by jobovy** [comment]
Fixed in #292.
